### PR TITLE
Add builder for creating TaskPool and add examples

### PR DESCRIPTION
**Pull request by aclysma**
Set up a builder for constructing the TaskPool with customizable thread count, stack size, and thread name. Add a couple examples to demonstrate busy and idle workloads.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -27,41 +27,109 @@ macro_rules! pin_mut {
...
+    num_threads: Option<usize>,
+    /// If set, we'll use the given stack size rather than the system default
+    stack_size: Option<usize>,
+    /// Allows customizing the name of the threads - helpful for debugging
+    thread_name: Option<String>,
+}
+
+impl TaskPoolBuilder {
+    /// Creates a new TaskPoolBuilder instance
+    pub fn new() -> Self {
```
Comment (lachlansneff) [inline src/lib.rs:44]: My inclination is to instead make a `TaskPool::build` function, but I'll leave it up to you.

File: src/lib.rs [hunk 2]
```diff
@@ -27,41 +27,109 @@ macro_rules! pin_mut {
...
 
+/// Used to create a TaskPool
+#[derive(Debug, Default, Clone)]
+pub struct TaskPoolBuilder {
+    /// If set, we'll set up the thread pool to use at most n threads. Otherwise use
+    /// the logical core count of the system
+    num_threads: Option<usize>,
+    /// If set, we'll use the given stack size rather than the system default
+    stack_size: Option<usize>,
+    /// Allows customizing the name of the threads - helpful for debugging
```
Comment (lachlansneff) [inline src/lib.rs:38]: Note that the index gets appended to the name.

File: src/lib.rs [hunk 3]
```diff
@@ -27,41 +27,109 @@ macro_rules! pin_mut {
...
-    pub fn create_num_threads(num_threads: usize) -> Self {
+    pub(crate) fn new(
+        num_threads: Option<usize>,
+        stack_size: Option<usize>,
+        thread_name: Option<&str>,
+    ) -> Self {
         let executor = Arc::new(Executor::new());
         let shutdown_flag = Arc::new(AtomicBool::new(false));
 
+        let num_threads = if let Some(num_threads) = num_threads {
```
Comment (lachlansneff) [inline src/lib.rs:92]: Use `num_threads.unwrap_or_else`.

**Comment by aclysma** [comment]
Thanks, good comments all around!

**Comment by lachlansneff** [comment]
Fantastic, this looks great. Let's merge this.

**Pull request merged**